src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs

```diff
@@ -119,28 +119,34 @@ protected virtual void MessageHandling((IActorRef, object) args)
                         _buffer.Enqueue(consumerMessage);
                     
                     Pump();
-                    
                     break;
                 
                 case Status.Failure failure:
                     var exception = failure.Cause;
-                    switch (_decider(failure.Cause))
+                    if (_decider(exception) == Directive.Stop)
                     {
-                        case Directive.Stop:
-                            // Throw
-                            FailStage(exception);
-                            break;
-                        case Directive.Resume:
-                            // keep going
-                            break;
-                        case Directive.Restart:
-                            // TODO: Need to do something here: https://github.com/akkadotnet/Akka.Streams.Kafka/issues/33
-                            break;
-                    }
+                        FailStage(failure.Cause);
+                        break;
+                    } 
+                    
+                    var isSerializationError = exception is ConsumeException cEx && cEx.Error.IsSerializationError();
+                    if (isSerializationError)
+                        break;
+                    
+                    // Empty the buffer to make sure that messages does not get duplicated
+                    while (_buffer.TryDequeue(out _))
+                    { }
+                    
+                    // ConsumerActor are designed to suicide itself on any error except for de/serialization error
+                    // to prevent any offset/commit runaway. We will need to restart it.
+                    SourceActor.Unwatch(ConsumerActor);
+                    ConsumerActor = CreateConsumerActor();
+                    SourceActor.Watch(ConsumerActor);
+                    ConfigureSubscription();
                     break;
                 
                 case Terminated terminated:
-                    FailStage(new ConsumerFailed());
+                    Log.Info($"Consumer actor terminated: {terminated.ActorRef.Path}");
```

Comment:
So we don't terminate anymore when the consumer actor dies?

Reply:
No, we don't terminate the stage on consumer actor death, that will be handled by the decider.
The decider depends solely on the failure messages being passed from the actor and decides if it needs to re-create the dead actor or to fail the stage.

Reply:
And the read buffer is stored on the actor or inside the stage? I had thought it was the latter

Reply:
Buffer is held inside the stage, that gets purged when the stage is told to resume or restart.
The Kafka consumer lives inside the actor, that held the offset counter for the current session, that needs to be killed to make sure that we're restarting the offset on a resume or a restart.